src/Stratis.Bitcoin/P2P/PeerAddressManager.cs

```diff
@@ -0,0 +1,313 @@
+                        {
+                            connectTokenSource.CancelAfter(TimeSpan.FromSeconds(5));
+
+                            Node node = null;
+
+                            try
+                            {
+                                clonedParameters.ConnectCancellation = connectTokenSource.Token;
+
+                                var addressManagerBehaviour = clonedParameters.TemplateBehaviors.Find<PeerAddressManagerBehaviour>();
+                                clonedParameters.TemplateBehaviors.Clear();
+                                clonedParameters.TemplateBehaviors.Add(addressManagerBehaviour);
+
+                                node = Node.Connect(network, peer.Endpoint, clonedParameters);
+                                node.VersionHandshake(connectTokenSource.Token);
+                                node.MessageReceived += (s, a) =>
+                                {
+                                    if (a.Message.Payload is AddrPayload addressPayload)
+                                        Interlocked.Add(ref peersFound, addressPayload.Addresses.Length);
+                                };
+
+                                node.SendMessageAsync(new GetAddrPayload());
+
+                                connectTokenSource.Token.WaitHandle.WaitOne(2000);
+                            }
+                            catch
+                            {
+                            }
+                            finally
+                            {
+                                node?.DisconnectAsync();
+                            }
+                        }
+                    });
+                }
+                catch (OperationCanceledException)
+                {
+                    if (nodeParameters.ConnectCancellation.IsCancellationRequested)
+                        throw;
+                }
+            }
+
+            await Task.CompletedTask;
+        }
+
+        private void PopulateTableWithDNSNodes(Network network, List<NetworkAddress> peers)
+        {
+            peers.AddRange(network.DNSSeeds.SelectMany(seed =>
+            {
+                try
+                {
+                    return seed.GetAddressNodes();
+                }
+                catch (Exception)
+                {
+                    return new IPAddress[0];
+                }
+            })
+            .Select(d => new NetworkAddress(d, network.DefaultPort)));
+        }
+
+        private void PopulateTableWithHardNodes(Network network, List<NetworkAddress> peers)
+        {
+            peers.AddRange(network.SeedNodes);
+        }
+
+        public void Dispose()
+        {
+            this.asyncLoop?.Dispose();
+
+            SavePeers();
+        }
+    }
+
+    public static class PeerAddressExtensions
+    {
+        public static IEnumerable<PeerAddress> New(this ConcurrentDictionary<int, PeerAddress> peers)
```

Comment:
Consider using .Skip(0) on ConcurrentDictionary iterations to avoid holding a lock.

Reply:
Fixed